Update protein_coding validator and add unit tests

SchistoDan opened the pull request:
- Updated protein_coding.py with new hmm alignment requirements (split seq on gaps, align individually, recombine)
- Added test_protein_coding_validator.py (18 unit tests)
- Added test data in exmaples/

rvosa commented:
Hi @SchistoDan,

I see that, unfortunately, some of the unit tests are failing. Could you address those before we merge?

Thanks!

Rutger

SchistoDan commented:
Hi @rvosa

It looks like 3 tests in `test_structval.py` are failing due to changes I made to `protein_coding.py` - the errors seem to show that 3 methods were expecting stop codons in the test sequences for that test module.

I'll update test_structval.py and re-test.

I've also been doing some additional testing of protein_coding.py, and I think some hmm alignment quality filtering needs to occur, as splitting the sequence into fragments can result in unexpected alignments (e.g. a 15bp fragment in the middle of the consensus may align at the end with very low posterior probability) - I'm in the process of adding a hmm alignment quality chekc method to checks that alignment coverage is good (>80% of original fragment aligned) and that >50% of positions have good PP scores (>7).

Hopefully I'll get these all to you tomorrow. Apologies for the delay

Cheers,
Dan

SchistoDan commented:
Hi @rvosa!

After several attempts, all the unit tests, linting, etc. pass. 

I ended up changing the hmm logic after the testing I mentioned in my last comment, as follows:

1. Remove tildes: ATG~CCC---AAA~TTTGGG → ATGCCC---AAATTTGGG
2. Split on gaps: Create fragments [ATGCCC] and [AAATTTGGG]
3. Multi-FASTA nhmmer: Run nhmmer on all fragments together
4. Parse HMM cooordinates for each fragment
5. Place fragments at correct HMM positions (i.e. in HMM coordinate space)
6. Analysis: Reading frames, stop codons, etc.

This logic/process is implemented in my current pull request [protein_coding.py and test_protein_coding_validator.py updates 2](https://github.com/naturalis/barcode_validator/pull/62/commits/58ea2d380126d364281e134bc72e591257fd2a25). 

However, after going through the logic, logs and outputs, and chatting to Ben today, we noticed a few issues, regarding gaps, and so think the logic needs to be updated slightly to:

1. removed tildes
2. Replaced gaps ('-') with Ns
3. run nhmmer
4. any leading/trailing Ns/gaps removed
5. internal Ns kept in sequence

I think this is necessary as splitting fragments on gaps and then aligning them to the HMM sometimes resulted in a different number of gaps being inserted between HMM-aligned fragments in the output in comparison to the original sequence, e.g.:
>original seq
frag1-[213 gaps]-frag2
>HMM-aligned barcode region seq
frag1-[210 gaps]-frag2

By replacing the gaps with Ns, and therefore avoiding splitting consensus sequences into fragments, the gap structure is maintained and the Ns are considered (seemingly as 'neutral') in the alignment. I'm in the process of making these changes and will make a new commit to the PR when they're done - just wanted to keep you in the loop.

Let me know your thoughts. Happy to have a chat on this if you'd like, and can expand on why I change hmmalign to nhmmer.

Best,
Dan
